pkg/reconciler/pipelinerun/resources/apply.go

```diff
@@ -150,13 +150,70 @@ func ApplyContexts(spec *v1.PipelineSpec, pipelineName string, pr *v1.PipelineRu
 	return ApplyReplacements(spec, GetContextReplacements(pipelineName, pr), map[string][]string{}, map[string]map[string]string{})
 }
 
+// filterMatrixContextVar returns a list of params which contain any matrix context variables such as
+// $(tasks.<pipelineTaskName>.matrix.length) and $(tasks.<pipelineTaskName>.matrix.<resultName>.length)
+func filterMatrixContextVar(params v1.Params) v1.Params {
+	var filteredParams v1.Params
+	for _, param := range params {
+		if expressions, ok := param.GetVarSubstitutionExpressions(); ok {
+			for _, expression := range expressions {
+				// tasks.<pipelineTaskName>.matrix.length
+				// tasks.<pipelineTaskName>.matrix.<resultName>.length
+				subExpressions := strings.Split(expression, ".")
+				if subExpressions[2] == "matrix" && subExpressions[len(subExpressions)-1] == "length" {
+					filteredParams = append(filteredParams, param)
+				}
+			}
+		}
+	}
+	return filteredParams
+}
+
 // ApplyPipelineTaskContexts applies the substitution from $(context.pipelineTask.*) with the specified values.
-// Uses "0" as a default if a value is not available.
-func ApplyPipelineTaskContexts(pt *v1.PipelineTask) *v1.PipelineTask {
+// Uses "0" as a default if a value is not available as well as matrix context variables
+// $(tasks.<pipelineTaskName>.matrix.length) and $(tasks.<pipelineTaskName>.matrix.<resultName>.length)
+func ApplyPipelineTaskContexts(pt *v1.PipelineTask, pipelineRunStatus v1.PipelineRunStatus, facts *PipelineRunFacts) *v1.PipelineTask {
 	pt = pt.DeepCopy()
+	var pipelineTaskName string
+	var resultName string
+	var matrixLength int
+
 	replacements := map[string]string{
 		"context.pipelineTask.retries": strconv.Itoa(pt.Retries),
 	}
+
+	filteredParams := filterMatrixContextVar(pt.Params)
+
+	for _, p := range filteredParams {
+		pipelineTaskName, resultName = p.ParseTaskandResultName()
+		// find the referenced pipelineTask to count the matrix combinations
+		if pipelineTaskName != "" && pipelineRunStatus.PipelineSpec != nil {
+			for _, task := range pipelineRunStatus.PipelineSpec.Tasks {
+				if task.Name == pipelineTaskName {
+					matrixLength = task.Matrix.CountCombinations()
+				}
+			}
+			replacements["tasks."+pipelineTaskName+".matrix.length"] = strconv.Itoa(matrixLength)
+		}
+		// find the resultName from the ResultsCache
+		if pipelineTaskName != "" && resultName != "" {
+			var resultsLength int
+			for _, pt := range facts.State {
+				if pt.PipelineTask.Name == pipelineTaskName {
+					var resultsCache *map[string][]string
+					if len(pt.ResultsCache) == 0 {
+						cache := createResultsCacheMatrixedTaskRuns(pt)
+						resultsCache = &cache
+						pt.ResultsCache = *resultsCache
+					}
+					resultsLength = len(pt.ResultsCache[resultName])
+					continue
+				}
+			}
+			replacements["tasks."+pipelineTaskName+".matrix."+resultName+".length"] = strconv.Itoa(resultsLength)
+		}
```

Comment:
is this the same?
```suggestion
				if pt.PipelineTask.Name == pipelineTaskName {
					if len(pt.ResultsCache) == 0 {
						pt.ResultsCache = createResultsCacheMatrixedTaskRuns(pt)
					}
					replacements["tasks."+pipelineTaskName+".matrix."+resultName+".length"] = strconv.Itoa(len(pt.ResultsCache[resultName]))
					continue
				}
```

Reply:
The reason for assigning a pointer is to persist the cached results in the state. This way does not persist the cachedResults in that state.

Reply:
Is there a test to confirm that? I think if I change the code, the unit tests can still pass

Reply:
The test `TestReconciler_PipelineTaskMatrixExplicitCombosResultsAndMatrixContextVars` fails with p-dag-2 and p-dag3 as it fails to retrieve the context variables from the cache. https://github.com/tektoncd/pipeline/blob/542c3b36307b9713fe0897aba726df12cbf537b0/pkg/reconciler/pipelinerun/pipelinerun_test.go#L12715

Reply:
emmm it doesn't fail from my end. I just feel like there are some duplicate logic of this code and inside the createResultsCacheMatrixedTaskRuns and not sure if they are necessary

Reply:
Sorry, you're right. I tried to replace the code with yours before and it failed, but it must have been because of something else. Just updated!